pkg/pillar/canbus/canbus.go

```diff
@@ -0,0 +1,853 @@
+			nmsg, err := nl.ParseRouteAttr(datum.Value)
+			if err != nil {
+				continue
+			}
+			for _, nested := range nmsg {
+				switch nested.Attr.Type {
+				case IFLA_CAN_TDC_TDCV:
+					link.TDCV = native.Uint32(nested.Value)
+				case IFLA_CAN_TDC_TDCO:
+					link.TDCO = native.Uint32(nested.Value)
+				case IFLA_CAN_TDC_TDCF:
+					link.TDCF = native.Uint32(nested.Value)
+				}
+			}
+		}
+	}
+}
+
+// Set a bit flag
+func setBitFlag(flags uint32, mask uint32) uint32 {
+	return (flags | mask)
+}
+
+// Clear a bit flag
+func clearBitFlag(flags uint32, mask uint32) uint32 {
+	return (flags & (^mask))
+}
+
+// Convert Sample Point string (float format) value to uint32
+// (multiplying by 1000)
+func convSamplePoint(str string) (uint32, error) {
+	value, err := strconv.ParseFloat(str, 32)
+	if err != nil {
+		return 0, err
+	}
+	return uint32(value * 1000), nil
+}
+
+// Convert string value to uint32
+func convUint32(str string) (uint32, error) {
+	value, err := strconv.ParseUint(str, 0, 32)
+	if err != nil {
+		return 0, err
+	}
+	return uint32(value), nil
+}
+
+// Convert string value to uint16
+func convUint16(str string) (uint16, error) {
+	value, err := strconv.ParseUint(str, 0, 16)
+	if err != nil {
+		return 0, err
+	}
+	return uint16(value), nil
+}
+
+// Set the proper flag of unix.CANCtrlMode according to the state (e.g.
+// "on" or "off"), except the TDC mode
+func setCtrlModeFlags(flagBit uint32, state string, ctrlMode *unix.CANCtrlMode) error {
+	if ctrlMode == nil {
+		return errors.New("No unix.CANCtrlMode provided")
+	}
+	mask := ctrlMode.Mask
+	flag := ctrlMode.Flags
+
+	if state == "on" {
+		mask = setBitFlag(mask, flagBit)
+		flag = setBitFlag(flag, flagBit)
+	} else {
+		mask = clearBitFlag(mask, flagBit)
+		flag = clearBitFlag(flag, flagBit)
+	}
+
+	ctrlMode.Mask = mask
+	ctrlMode.Flags = flag
+	return nil
+}
+
+// Set TDC mode flags of unix.CANCtrlMode according to the state provided as
+// string
+func setCtrlModeTDC(state string, ctrlMode *unix.CANCtrlMode) error {
+	if ctrlMode == nil {
+		return errors.New("No unix.CANCtrlMode provided")
+	}
+	mask := ctrlMode.Mask
+	flag := ctrlMode.Flags
+
+	switch state {
+	case "auto":
+		mask = setBitFlag(mask, unix.CAN_CTRLMODE_TDC_AUTO)
+		flag = setBitFlag(flag, unix.CAN_CTRLMODE_TDC_AUTO)
+	case "manual":
+		mask = setBitFlag(mask, unix.CAN_CTRLMODE_TDC_MANUAL)
+		flag = setBitFlag(flag, unix.CAN_CTRLMODE_TDC_MANUAL)
+	case "off":
+		mask = setBitFlag(mask,
+			unix.CAN_CTRLMODE_TDC_AUTO|unix.CAN_CTRLMODE_TDC_MANUAL)
+		flag = setBitFlag(flag,
+			unix.CAN_CTRLMODE_TDC_AUTO|unix.CAN_CTRLMODE_TDC_MANUAL)
+	}
+
+	ctrlMode.Mask = mask
+	ctrlMode.Flags = flag
+	return nil
+}
+
+// Setup CAN interface parameters, return in case of any error
+// Parameters to be set are marked in the array props in the following order:
+// BitTiming, DataBitTiming, CtrlMode, Termination, Restart, RestartMs, TDC
+func setupCANParams(link *CAN, canProps CAN, props []bool) error {
```

Comment:
to reduce the complexity of SetupCAN function and make Yetus happy :smile:

Reply:
Yes, but no.

This is not the way to go; please don't use bool arrays with magic numbers.

What about doing the map with funcs instead?

Reply:
Is not easy to use map in this case due to the different types of the arguments. I would like to avoid create wrapper functions to force the map usage, but I have an alternative approach to avoid the array of bools... let's discuss it tomorrow...